src/trs/rewrite.rs

```diff
@@ -461,32 +363,46 @@ impl TRS {
             .expect("poisoned lexicon")
             .background
             .len();
-        if num_background >= num_rules - 1 {
-            return Ok(trs);
-        }
         let ref_idx: usize = rng.gen_range(0, num_rules);
-        let mut target_idx: usize = rng.gen_range(num_background, num_rules);
-        while ref_idx == target_idx {
-            target_idx = rng.gen_range(num_background, num_rules);
-        }
         let mut ref_rule = trs.utrs.rules[ref_idx].clone();
         let rhs_idx = rng.gen_range(0, ref_rule.rhs.len());
         let temp_rule = Rule::new(ref_rule.lhs, vec![ref_rule.rhs[rhs_idx].clone()]);
         if temp_rule == None {
-            return Ok(trs);
+            return Err(SampleError::OptionsExhausted);
         }
         ref_rule = temp_rule.unwrap();
-        if ref_rule.lhs.variables().len() != ref_rule.rhs[0].variables().len() {
-            return Ok(trs);
+        if ref_rule.lhs.variables().len() < ref_rule.rhs[0].variables().len() {
+            return Err(SampleError::OptionsExhausted);
+        }
+        Ok(ref_rule)
+    }
+    /// Selects two Rules from the TRS at random and atempts to inverse evaluate one rule on the other, if it
+    /// succeeds it takes that new rule and inserts it imediately after the background.
+    pub fn inverse_evaluate<R: Rng>(&self, rng: &mut R) -> Result<TRS, SampleError> {
+        let mut trs = self.clone();
+        let num_rules = self.len();
+        let num_background = self
+            .lex
+            .0
+            .read()
+            .expect("poisoned lexicon")
+            .background
+            .len();
+        if num_background >= num_rules - 1 {
+            return Err(SampleError::OptionsExhausted);
         }
+        let mut target_idx: usize = rng.gen_range(num_background, num_rules);
+        let ref_rule = trs.sample_rule(rng)?;
         let new_rule =
             TRS::inverse_evaluate_rule_helper(&ref_rule, &trs.utrs.rules[target_idx], rng);
         if new_rule == None {
-            return Ok(trs);
+            return Err(SampleError::OptionsExhausted);
         }
-        trs.utrs.remove_idx(target_idx).expect("removing old rule");
         trs.utrs
```

Comment:
I think you can use `?` here instead of `expect`. That way, you return the error rather than panic. Something like this:

```rust
trs.utrs.remove_clauses(&ref_rule)?;
```